### ISSUE TITLE
HTTPRepositoryConnection not public

### ISSUE THREAD

**Issue by woutermont** [eclipse/rdf4j#1519]
Is there a reason why [`HTTPRepositoryConnection`](https://github.com/eclipse/rdf4j/blob/master/core/repository/http/src/main/java/org/eclipse/rdf4j/repository/http/HTTPRepositoryConnection.java) is not public?

**Comment by jeenbroekstra** [comment]
Well, mainly because it doesn't need to be. An instance of the class is obtained by calling `HTTPRepository.getConnection`.

**Comment by woutermont** [comment]
That seems obvious, but isn't it a rather ad hoc argument? There are tons of classes that do not _need_ to be public but still are (in fact it's rather hard to find many private classes in the rdf4j packages). Moreover, since the class is (package) private, one cannot access its session, nor extend it to circumvent this. I stumbled upon this when trying to access the session in order to debug a subtle problem in a project of mine (for which in the end I simply forked rdf4j to gain access). Fair enough answer, however, so feel free to close this issue.

**Comment by jeenbroekstra** [comment]
> That seems obvious, but isn't it a rather ad hoc argument? There are tons of classes that do not _need_ to be public but still are (in fact it's rather hard to find many private classes in the rdf4j packages).

You're right that this is not something we are as consistent with as we should be. We use a principle of making things as private as possible unless there's a good reason to open it up. Over the years, we have been more diligent in doing this right in some places than in others. The trouble is, of course, that once you have opened something up as public, you can't turn that back until a major release comes along, because it is a backward-compatibility-breaking change.

In general, (package-)private classes in rdf4j are marked as such because we _don't_ want them to be considered part of the public API, and discourage people from relying on their internal details as we may change those details between releases, without warning.

> Moreover, since the class is (package) private, one cannot access its session, nor extend it to circumvent this. I stumbled upon this when trying to access the session in order to debug a subtle problem in a project of mine (for which in the end I simply forked rdf4j to gain access). Fair enough answer, however, so feel free to close this issue.

There are ways to solve this other than making the class public. I just had a quick look and I am left wondering why HTTPRepositoryConnection doesn't extend SessionManagerDependent. I'll keep this issue open to have a closer look at that.
